File: src/main/scala/MyTest/mmlTests/a201401b/Scala/gpt35turbo/ProgressiveAcceptorImpl.scala
package MyTest.mmlTests.a201401b.Scala.gpt35turbo

import MyTest.mmlTests.a201401b.Scala.Aggregator;
import MyTest.mmlTests.a201401b.Scala.ProgressiveAcceptor;
import MyTest.mmlTests.a201401b.Scala.ProgressiveFilter;

class ProgressiveAcceptorImpl[X] extends ProgressiveAcceptor[X] {

  private var filter: ProgressiveFilter[X] = _
  private var aggregator: Aggregator[X] = _
  private var size: Int = _
  private var sequence: List[X] = Nil

  override def setProgressiveFilter(filter: ProgressiveFilter[X]): Unit = {
    this.filter = filter
  }

  override def setAggregator(aggregator: Aggregator[X]): Unit = {
    this.aggregator = aggregator
  }

  override def setSize(size: Int): Unit = {
    if (size < 0) {
      throw new IllegalArgumentException("Not negative!")
    }
    this.size = size
  }

  override def accept(pos: Int, elem: X): Boolean = {
    if (filter == null || aggregator == null || size == 0) {
      throw new IllegalStateException("Not correctly initialized")
    }
    if (pos < sequence.length) {
      // Remove elements after the position
      sequence = sequence.take(pos)
    }
    if (pos != sequence.length) {
      throw new IllegalArgumentException("Wrong position")
    }
    if (pos == size) {
      false
    } else {
      val previous = sequence.headOption.getOrElse(elem)
      if (filter.isNextOK(previous, elem)) {
        sequence = elem :: sequence
        true
      } else {
        false
      }
    }
  }

  override def aggregateAll(): X = {
    if (aggregator == null) {
      throw new NullPointerException("Aggregator not set")
    }
    sequence.reduceOption(aggregator.aggregate).getOrElse(throw new IllegalStateException("No elements"))
  }
}

File: src/main/scala/mmlTests/a201401b/java/gpt35turbo/ProgressiveAcceptorImpl.java
package mmlTests.a201401b.java.gpt35turbo;

import java.util.ArrayList;
import java.util.List;

import mmlTests.a201401b.java.Aggregator;
import mmlTests.a201401b.java.ProgressiveAcceptor;
import mmlTests.a201401b.java.ProgressiveFilter;

public class ProgressiveAcceptorImpl<X> implements ProgressiveAcceptor<X> {

    private ProgressiveFilter<X> filter;
    private Aggregator<X> aggregator;
    private int size;
    private List<X> sequence;

    public ProgressiveAcceptorImpl() {
        sequence = new ArrayList<>();
    }

    @Override
    public void setProgressiveFilter(ProgressiveFilter<X> filter) {
        this.filter = filter;
    }

    @Override
    public void setAggregator(Aggregator<X> aggregator) {
        this.aggregator = aggregator;
    }

    @Override
    public void setSize(int size) {
        this.size = size;
    }

    @Override
    public boolean accept(int pos, X elem) {
        if (filter == null || size == 0) {
            throw new IllegalStateException();
        }
        if (pos > sequence.size()) {
            throw new IndexOutOfBoundsException();
        }
        if (pos < sequence.size()) {
            sequence.subList(pos, sequence.size()).clear();
        }
        if (pos >= sequence.size()) {
            sequence.add(elem);
        } else {
            sequence.set(pos, elem);
        }
        return filter.isNextOK(sequence.get(pos - 1), elem);
    }

    @Override
    public X aggregateAll() {
        X aggregate = null;
        for (X elem : sequence) {
            if (aggregate == null) {
                aggregate = elem;
            } else {
                aggregate = aggregator.aggregate(aggregate, elem);
            }
        }
        return aggregate;
    }

}
